tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,153 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.Stackable
+
+import scala.collection.immutable.HashMap
+
+/**
+ * An abstract cache that assigns `TargetT`-typed values to `SourceT` typed values, while potentially modifying a
+ * `ContextT`-typed context.
+ *
+ * Example: IntValueCache assigns [[ArenaCell]]s to integers, while modifying a [[PureArena]].
+ *
+ * While it extends [[Stackable]], there's no need to implement an actual stack data structure. By assigning each cache
+ * entry a level, effectively a timestamp for when it was added, we can proxy the state of the stack, by filtering out
+ * entries, based on their levels.
```

Comment:
iiuc, this is a comment for the implementation but not needed for documenting the class? Maybe move it into a normal comment near the hashmaps that implement this approach?